File: Data Structures & Algorithms/duplicate-integer/submission-1.cs
public class Solution {
    public bool hasDuplicate(int[] nums) {
        // Use a HashSet to store distinct nums in while checking if it contains the current num
        // Time Complexity is O(n) where n is the lenght of nums
        // Space Complexity is O(n) where is n also the length of nums

        var hashSet = new HashSet<int>();

        foreach (var num in nums) {
            if(hashSet.Contains(num)) return true;
            hashSet.Add(num);
        }

        return false;
    }
}

File: Data Structures & Algorithms/subsets/submission-1.ts
// we can build a recursive decision tree where we have to choices/nodes at every branch. 
// The left will include the current num, and the right will exclude it. 
// This will ensure that we have no duplicate subsets.
// Our base case to return from the current recursive call would be if the current index equal to the length of the nums arr

// Time Complexity would be O(n * 2^n). n - because we will be creating a copy of the current subset before add
// it to the output. 2^n because we have two choices at every branch
// Space Complexity would O(n) (NOT SURE WHY)

class Solution {
    /**
     * @param {number[]} nums
     * @return {number[][]}
     */
    subsets(nums: number[]): number[][] {
        let res = []

        const dfs = (i: number, subset: number[]) => {
            // base case
            if (i === nums.length) {
                res.push([...subset]) // Time: O(n)
                return 
            }

            // left. include curr num
            // [1] 0
            // [1, 2] 1
            // [1, 2, 3] 2
            // [1, 2, 3] 3
            subset.push(nums[i])
            dfs(i + 1, subset)

            // right. exclude curr num
            // [1, 2] 3
            // [1] 2
            subset.pop()
            dfs(i + 1, subset)
        }

        dfs(0, [])
        return res
    }
}
